diff --git a/src/extra/blas/Makefile.in b/src/extra/blas/Makefile.in
index 3661416..c9b53e2 100644
--- a/src/extra/blas/Makefile.in
+++ b/src/extra/blas/Makefile.in
@@ -17,7 +17,7 @@ include $(top_builddir)/Makeconf
 
 ALL_CFLAGS = $(ALL_CFLAGS_LO)
 ALL_FFLAGS = $(ALL_FFLAGS_LO)
-ALL_FCLAGS = $(ALL_FFLAGS_LO)
+ALL_FCFLAGS = $(ALL_FFLAGS_LO)
 
 SOURCES = blas00.c blas.f cmplxblas.f blas2.f90 cmplxblas2.f90
 
diff --git a/src/modules/lapack/Makefile.in b/src/modules/lapack/Makefile.in
index 8e593ea..e35c956 100644
--- a/src/modules/lapack/Makefile.in
+++ b/src/modules/lapack/Makefile.in
@@ -15,6 +15,10 @@ include $(top_builddir)/Makeconf
 .f90.o:
 	$(FC) $(ALL_FCFLAGS) -c @FCFLAGS_f90@ $< -o $@
 
+ALL_CFLAGS = $(ALL_CFLAGS_LO)
+ALL_FFLAGS = $(ALL_FFLAGS_LO)
+ALL_FCFLAGS = $(ALL_FFLAGS_LO)
+
 SOURCES_C = Lapack.c @USE_VECLIB_G95FIX_TRUE@ vecLibg95c.c
 SOURCES_F = @USE_VECLIB_G95FIX_TRUE@ vecLibg95f.f
 
