diff --git a/src/main/devices.c b/src/main/devices.c
index 850f9fa..89baeb3 100644
--- a/src/main/devices.c
+++ b/src/main/devices.c
@@ -420,7 +420,7 @@ double defaultStrWidth(const char *str, const pGEcontext gc, pDevDesc dd)
 { return 0; }
 void noopText(double x, double y, const char *str, double rot,
               double hadj, const pGEcontext gc, pDevDesc dd) {}
-SEXP defaultGetEvent(SEXP, const char *)
+SEXP defaultGetEvent(SEXP ev, const char * x)
 { return R_NilValue; }
 void noopTextUTF8(double x, double y, const char *str, double rot,
                   double hadj, const pGEcontext gc, pDevDesc dd) {}
